jerry-core/ecma/base/ecma-helpers-value.c

```diff
@@ -33,60 +30,87 @@ JERRY_STATIC_ASSERT (sizeof (ecma_value_t) * JERRY_BITSINBYTE >= ECMA_VALUE_SIZE
-                                              ECMA_VALUE_TYPE_WIDTH);
+  return value & ECMA_VALUE_TYPE_MASK;
 } /* ecma_get_value_type_field */
 
 /**
- * Get value field of ecma value
+ * Convert a pointer into an ecma value.
  *
- * @return value field
+ * @return ecma value
  */
-static uintptr_t __attr_pure___
-ecma_get_value_value_field (ecma_value_t value) /**< ecma value */
+static inline ecma_value_t __attr_pure___ __attr_always_inline___
+ecma_pointer_to_ecma_value (const void *ptr) /**< pointer */
 {
-  return (uintptr_t) JRT_EXTRACT_BIT_FIELD (ecma_value_t, value,
-                                            ECMA_VALUE_VALUE_POS,
-                                            ECMA_VALUE_VALUE_WIDTH);
-} /* ecma_get_value_value_field */
+  if (sizeof (void *) <= sizeof (ecma_value_t))
+  {
+    uintptr_t uint_ptr = (uintptr_t) ptr;
+    JERRY_ASSERT ((uint_ptr & ECMA_VALUE_FULL_MASK) == 0);
+    return (ecma_value_t) uint_ptr;
+  }
+  else
+  {
+    mem_cpointer_t ptr_cp;
+    ECMA_SET_NON_NULL_POINTER (ptr_cp, ptr);
+
+    return ((ecma_value_t) ptr_cp) << ECMA_VALUE_SHIFT;
+  }
+} /* ecma_pointer_to_ecma_value */
 
 /**
- * Set type field of ecma value
+ * Get a pointer from an ecma value
  *
- * @return ecma value with updated field
+ * @return pointer
  */
-static ecma_value_t __attr_pure___
-ecma_set_value_type_field (ecma_value_t value, /**< ecma value to set field in */
-                           ecma_type_t type_field) /**< new field value */
+static inline void * __attr_pure___ __attr_always_inline___
+ecma_get_pointer_from_ecma_value (ecma_value_t value) /**< value */
 {
-  return JRT_SET_BIT_FIELD_VALUE (ecma_value_t, value,
-                                  type_field,
-                                  ECMA_VALUE_TYPE_POS,
-                                  ECMA_VALUE_TYPE_WIDTH);
-} /* ecma_set_value_type_field */
+  if (sizeof (void *) <= sizeof (ecma_value_t))
+  {
+    return (void *) (uintptr_t) ((value) & ~ECMA_VALUE_FULL_MASK);
+  }
+  else
+  {
+    return ECMA_GET_NON_NULL_POINTER (ecma_number_t,
+                                      value >> ECMA_VALUE_SHIFT);
+  }
+} /* ecma_get_pointer_from_ecma_value */
+
+#define ECMA_SIMPLE_VALUE_TO_VALUE(simple_value) \
+  ((((ecma_value_t) (simple_value)) << ECMA_VALUE_SHIFT) | ECMA_TYPE_SIMPLE)
 
 /**
- * Set value field of ecma value
+ * Check whether the value is a given simple value.
  *
- * @return ecma value with updated field
+ * @return true - if the value is equal to the given simple value,
+ *         false - otherwise.
  */
-static ecma_value_t __attr_pure___
-ecma_set_value_value_field (ecma_value_t value, /**< ecma value to set field in */
-                            uintptr_t value_field) /**< new field value */
+static inline bool __attr_pure___ __attr_always_inline___
+ecma_is_value_equal_to_simple_value (ecma_value_t value, /**< ecma value */
+                                     ecma_simple_value_t simple_value) /**< simple value */
```

Comment:
The name is misleading. This is type of simple_value, so I suggest "value_type" or "simple_value_type"